bip-0360.mediawiki

```diff
@@ -0,0 +1,608 @@
+[https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki#bech32 BIP-173]. Bech32 encoding maps version 3 to the
+prefix <code>r</code>.
+
+Example P2QRH address:
+
+<code>bc1r...</code> (32-byte Bech32m-encoded HASH256 of the HASH256 of the public keys)
+
+=== ScriptPubKey ===
+
+The <code>scriptPubKey</code> for a P2QRH output is:
+
+  OP_PUSHNUM_3 OP_PUSHBYTES_32 <nowiki><hash></nowiki>
+
+Where:
+
+* <code>OP_PUSHNUM_3</code> (<code>0x03</code>) indicates SegWit version 3.
+* <nowiki><hash></nowiki> is the 32-byte HASH256 of the concatenated HASH256 of each public key.
+
+==== Hash Computation ====
+
+The hash is computed as a merkle tree of public key hashes:
+
+1. For each public key, compute its HASH256
+2. Pair the hashes and compute HASH256 of their concatenation
+3. Continue pairing and hashing until reaching a single root hash
+4. The final hash is the merkle root
+
+For example with 4 public keys:
+
+  h1 = HASH256(pubkey1)
+  h2 = HASH256(pubkey2)
+  h3 = HASH256(pubkey3)
+  h4 = HASH256(pubkey4)
+
+  h12 = HASH256(h1 <nowiki>||</nowiki> h2)
+  h34 = HASH256(h3 <nowiki>||</nowiki> h4)
+
+  root = HASH256(h12 <nowiki>||</nowiki> h34)
+
+When spending, if a public key hash is provided in the attestation with an empty signature, that hash will be used
+directly in the merkle tree computation rather than hashing the full public key. This allows excluding unused public
+keys from the transaction while still proving they were part of the original commitment.
+
+This merkle tree construction creates an efficient cryptographic commitment to multiple public keys while enabling
+selective disclosure.
+
+This allows for inclusion of a Taproot MAST merkle root in the attestation, which makes P2QRH a quantum-resistant
+version of Taproot.
+
+=== Transaction Serialization ===
+
+Following BIP-141, a new transaction serialization format is introduced to include an attestation field after the witness field:
+
+  [nVersion][marker][flag][txins][txouts][witness][attestation][nLockTime]
+
+* <code>marker</code>: <code>0x00</code> (same as SegWit)
+
+* <code>flag</code>: <code>0x02</code> (indicates the presence of both witness and attestation data)
+
+* <code>attestation</code>: Contains the quantum-resistant public keys and signatures.
+
+=== Attestation Structure ===
+
+The attestation field consists of:
+
+* <code>num_pubkeys</code>: The number of public keys
+([https://learnmeabitcoin.com/technical/general/compact-size/ compact size]).
+
+For each public key:
+
+* <code>pubkey_length</code>: compact size length of the public key.
+* <code>pubkey</code>: The public key bytes.
+
+Then:
+
+* <code>num_signatures</code>: The number of signatures (compact size).
+
+For each signature:
+
+* <code>signature_length</code>: compact size length of the signature.
+* <code>signature</code>: The signature bytes.
+
+This structure repeats for each input, in order, for flexibility in supporting multisig schemes and various
+quantum-resistant algorithms.
```

Comment:
I haven’t thought a lot about this, but given the goal of extensibility, it might be good to add a byte to indicate a signature type for more flexibility?

Reply:
As I go into in the `Signature Algorithm Identification` section, we just use the length of the key and signature to indicate signature type. If there's overlap, an extra byte is added.